File: api/moo.lua
local ffi = require("ffi")

ffi.cdef[[
    double absolute(double x);
    int64_t idiv(double a, double b);
    int64_t dfloor(double a);
    int64_t dceil(double a);
    int64_t gcd(int64_t a, int64_t b);
    int64_t lcm(int64_t a, int64_t b);
    int64_t mod(int64_t a, int64_t b);
    double flmod(double a, double b);
    int64_t fac(int64_t n);
    int64_t binom(int64_t n, int64_t k);
    int64_t* factable(int64_t n);
    void clearptr(int64_t* ptr);
    int64_t permutation(int n, int k);
    int64_t combination(int n, int k);
    int64_t* genPerm(int64_t n);
    void freePerm(int64_t* ptr);
    double max(double a, double b);
    double min(double a, double b);
]]

local lib = ffi.load("moo")

local moo = {}

function moo.abs(x)
    return lib.absolute(x)
end

function moo.idiv(a, b)
    return lib.idiv(a, b)
end

function moo.floor(a)
    return lib.dfloor(a)
end

function moo.ceil(a)
    return lib.dceil(a)
end

function moo.gcd(a, b)
    return lib.gcd(a, b)
end

function moo.lcm(a, b)
    return lib.lcm(a, b)
end

function moo.mod(a, b)
    return lib.mod(a, b)
end

function moo.flmod(a, b)
    return lib.flmod(a, b)
end

function moo.fac(n)
    return lib.fac(n)
end

function moo.binom(n, k)
    return lib.binom(n, k)
end

function moo.factable(n)
    if n <= 0 then error("n cannot be 0.") end
    local ptr = lib.factable(n)
    if ptr == nil then error("NULL pointer returned from factable function.") end
    local result = {}
    for i = 0, n do
        result[#result+1] = tonumber(ptr[i])
    end
    lib.clearptr(ptr)
    return result
end

function moo.permutation(n, k)
    return lib.permutation(n, k)
end

function moo.combination(n, k)
    return lib.combination(n, k)
end

function moo.genPerm(n)
    if n <= 0 then error("n cannot be 0.") end
    local ptr = lib.genPerm(n)
    if ptr == nil then error("NULL pointer returned from genPerm function.") end
    local total = 1
    for i = 1, n do total = total * i end
    local result = {}
    for i = 0, total-1 do
        local perm = {}
        for j = 0, n-1 do
            perm[#perm+1] = tonumber(ptr[i * n + j])
        end
        result[#result+1] = perm
    end
    lib.freePerm(ptr)
    return result
end

function moo.max(a, b)
    return lib.max(a, b)
end

function moo.min(a, b)
    return lib.min(a, b)
end

return moo

File: api/java/moo.java
import com.sun.jna.*;
import com.sun.jna.ptr.PointerByReference;
import java.util.ArrayList;
import java.util.List;

public class Moo {
    public interface MooLib extends Library {
        MooLib INSTANCE = Native.load("moo", MooLib.class);

        double absolute(double x);
        long idiv(double a, double b);
        long floor(double a);
        long ceil(double a);
        long gcd(long a, long b);
        long lcm(long a, long b);
        long mod(long a, long b);
        double flmod(double a, double b);
        long fac(long n);
        long binom(long n, long k);
        Pointer factable(long n);
        long permutation(int n, int k);
        long combination(int n, int k);
        Pointer genPerm(long n);
        double max(double a, double b);
        double min(double a, double b);
    }

    public static double abs(double x) {
        return MooLib.INSTANCE.absolute(x);
    }

    public static long idiv(double a, double b) {
        return MooLib.INSTANCE.idiv(a, b);
    }

    public static long floor(double a) {
        return MooLib.INSTANCE.floor(a);
    }

    public static long ceil(double a) {
        return MooLib.INSTANCE.ceil(a);
    }

    public static long gcd(long a, long b) {
        return MooLib.INSTANCE.gcd(a, b);
    }

    public static long lcm(long a, long b) {
        return MooLib.INSTANCE.lcm(a, b);
    }

    public static long mod(long a, long b) {
        return MooLib.INSTANCE.mod(a, b);
    }

    public static double flmod(double a, double b) {
        return MooLib.INSTANCE.flmod(a, b);
    }

    public static long fac(long n) {
        return MooLib.INSTANCE.fac(n);
    }

    public static long binom(long n, long k) {
        return MooLib.INSTANCE.binom(n, k);
    }

    public static long[] factable(long n) {
        if (n <= 0) throw new IllegalArgumentException("n cannot be 0.");
        Pointer ptr = MooLib.INSTANCE.factable(n);
        if (ptr == null) throw new RuntimeException("NULL pointer returned from factable function.");
        long[] result = ptr.getLongArray(0, (int) n + 1);
        MooLib.INSTANCE.clearptr(ptr);
        return result;
    }

    public static long permutation(int n, int k) {
        return MooLib.INSTANCE.permutation(n, k);
    }

    public static long combination(int n, int k) {
        return MooLib.INSTANCE.combination(n, k);
    }

    public static List<long[]> genPerm(int n) {
        if (n <= 0) throw new IllegalArgumentException("n cannot be 0.");
        Pointer ptr = MooLib.INSTANCE.genPerm(n);
        if (ptr == null) throw new RuntimeException("NULL pointer returned from genPerm function.");
        int total = 1;
        for (int i = 1; i <= n; i++) total *= i;
        List<long[]> result = new ArrayList<>();
        for (int i = 0; i < total; i++) {
            long[] perm = new long[n];
            for (int j = 0; j < n; j++) {
                perm[j] = ptr.getLong((i * n + j) * Native.getNativeSize(Long.TYPE));
            }
            result.add(perm);
        }
        MooLib.INSTANCE.freeptr(ptr);
        return result;
    }

    public static double max(double a, double b) {
        return MooLib.INSTANCE.max(a, b);
    }

    public static double min(double a, double b) {
        return MooLib.INSTANCE.min(a, b);
    }
}